mbtiles.go

```diff
@@ -0,0 +1,292 @@
+	PNG:  "png",
+	JPG:  "jpg",
+	PBF:  "pbf",
+	WEBP: "webp",
+}
+
+var TileContentType = map[TileFormat]string{
+	PNG:  "image/png",
+	JPG:  "image/jpeg",
+	PBF:  "application/x-protobuf", // Content-Encoding header must be gzip
+	WEBP: "image/webp",
+}
+
+type Mbtiles struct {
+	filename           string
+	db                 *sql.DB
+	metadata           map[string]interface{}
+	tileformat         TileFormat // tile format: PNG, JPG, PBF
+	timestamp          time.Time  // timestamp of file, for cache control headers
+	hasUTFGrid         bool
+	utfgridCompression TileFormat
+	hasUTFGridData     bool
+}
+
+// Creates a new Mbtiles instance.
+// Connection is closed by runtime on application termination or by calling .Close() method.
+func NewMbtiles(filename string) (*Mbtiles, error) {
+	_, id := filepath.Split(filename)
+	id = strings.Split(id, ".")[0]
+
+	db, err := sql.Open("sqlite3", filename)
+	if err != nil {
+		return nil, err
+	}
+
+	//Saves last modified mbtiles time for setting Last-Modified header
+	fileStat, err := os.Stat(filename)
+	if err != nil {
+		log.Errorf("could not read file stats for mbtiles file: %s\n", filename)
+		return nil, err
+	}
+
+	//query a sample tile to determine format
+	var data []byte
+	err = db.QueryRow("select tile_data from tiles limit 1").Scan(&data)
+	if err != nil {
+		return nil, err
+	}
+	tileformat, err := detectTileFormat(&data)
+	if err != nil {
+		return nil, err
+	}
+	if tileformat == GZIP {
+		tileformat = PBF // GZIP masks PBF, which is only expected type for tiles in GZIP format
+	}
+	out := Mbtiles{
+		db:         db,
+		tileformat: tileformat,
+		timestamp:  fileStat.ModTime().Round(time.Second), // round to nearest second
+	}
+
+	// UTFGrids
+	// first check to see if requisite tables exist
+	var count int
+	err = db.QueryRow("SELECT count(*) FROM sqlite_master WHERE type='view' AND name = 'grids'").Scan(&count)
+	if err != nil {
+		log.Error(err)
+		return nil, err
+	}
+	if count == 1 {
+		// query a sample grid to detect type
+		var gridData []byte
+		err = db.QueryRow("select grid from grids where grid is not null LIMIT 1").Scan(&gridData)
+		if err != nil {
+			if err != sql.ErrNoRows {
+				log.Error("Could not read sample grid to determine type")
+				return nil, err
+			}
+		} else {
+			out.hasUTFGrid = true
+			out.utfgridCompression, err = detectTileFormat(&gridData)
+			if err != nil {
+				log.Error("Could not determine UTF Grid compression type")
+				return nil, err
+			}
+
+			// Check to see if grid_data view exists
+			count = 0 // prevent use of prior value
+			err = db.QueryRow("SELECT count(*) FROM sqlite_master WHERE type='view' AND name = 'grid_data'").Scan(&count)
+			if err != nil {
+				return nil, err
+			}
+			if count == 1 {
+				out.hasUTFGridData = true
+			}
+		}
+	}
+
+	return &out, nil
+
+}
+
+// Reads a grid at z, x, y into provided *[]byte.
+func (tileset *Mbtiles) ReadTile(z uint8, x uint64, y uint64, data *[]byte) error {
```

Comment:
@ka7eh this is a [method receiver](http://www.golangbootcamp.com/book/methods#sec-method_receivers); the syntax is a little unusual coming from other languages, but it is easiest to think about this as a method on an `Mbtiles` struct (aka object-like thing).  I adopted this pattern to better encapsulate behavior.

I'm still not completely sold on passing in the data slice pointer (`data *[]byte`) here for modification within this function; however, I followed the standard approach of Go's `Read` interface and the SQL `Scan` interface where the caller "owns" the byte slice, and passes it in to other functions for modification.   (note: "slice" is sort of but not quite like an array, but for purposes here I don't know that the distinction matters).